### PR TITLE
Dala 4294 api access for new roles

### PR DESCRIPTION
**Pull request by emanueldfine**
# Pull Request \<Title>
`<Description here>`
## Things to do during Peer Review
Please check all boxes before the Pull Request is merged. In case you skip a box, describe in the PRs description (that means: here) why the check is skipped.
- [x] The Github Actions (including Sonarqube Gateway and Lint Checks) are green. This is enforced by Github. 
- [x] A peer-review has been executed
  - [x] The code has been manually inspected by someone who did not implement the feature
  - [x] If this PR includes work on the frontend, at least one `@ts-nocheck` is removed. Additionally, there should not be any `@ts-nocheck` in files modified by this PR. If no `@ts-nocheck` are left: Celebrate :tada: :confetti_ball: type-safety and remove this entry. 
- [x] The PR actually implements what is described in the JIRA-Issue
- [x] At least one test exists testing the new feature
  - [x] If you have created new test files, make sure that they are included in a test container and actually run in the CI
- [x] At least 3 consecutive CI runs are successfully executed to ensure that there are no flaky tests.
- [x] Documentation is updated as required
- [x] The automated deployment is updated if required
- [x] If there was a database entity class added, there must also be a migration script for creating the corresponding database if flyway is already used by the service
- [ ] IF there are changes done to the framework data models or to a database entity class, the following steps were completed in order
  - [ ] A fresh clone of dataland.com is generated (see Wiki page on "OTC" for details)
  - [ ] The feature branch is deployed to clone with `Reset non-user related Docker Volumes & Re-populate` turned off
  - [ ] It's verified that the CD run is green
  - [ ] The new feature is manually used/tested/observed on the clone server. Testing of the new feature should (also) be done by a second, independent reviewer from the dev team
  - [ ] The feature branch is deployed to dev1 with `Reset non-user related Docker Volumes & Re-populate` turned on, and it's verified that the CD run is green  
- [x] ELSE, the new version is deployed to the dev server "dev1" using this branch
  - [x] Run with setting `Reset non-user related Docker Volumes & Re-populate` turned on 
  - [x] It's verified that this version actually is the one deployed (check gitinfo for branch name and commit id!)
  - [x] It's verified that the CD run is green
  - [x] The new feature is manually used/tested/observed on dev server. Testing of the new feature should (also) be done by a second, independent reviewer from the dev team

### CODE REVIEW

File: dataland-backend/src/main/kotlin/org/dataland/datalandbackend/api/DataDeletionApi.kt [hunk 1]
```diff
@@ -34,7 +34,17 @@ fun interface AdminDataManipulationApi {
         ],
     )
     @DeleteMapping()
-    @PreAuthorize("hasRole('ROLE_ADMIN')")
+    @PreAuthorize(
```
Comment (emanueldfine) [inline dataland-backend/src/main/kotlin/org/dataland/datalandbackend/api/DataDeletionApi.kt:37]: Reminder für uns:  Fragen ob Company-Data-Uploader nur ihre eigenen datasets löschen können sollen

File: dataland-backend/src/main/kotlin/org/dataland/datalandbackend/services/CompanyRoleChecker.kt [hunk 2]
```diff
@@ -17,23 +17,23 @@ import kotlin.reflect.full.memberProperties
...
-     * Method to check whether the currently authenticated user is company owner of a specified company and therefore
-     * has uploader rights for this company
+     * Checks whether the currently authenticated user has a company role for a specified company.
      * @param companyId the ID of the company
-     * @return a Boolean indicating whether the user is company owner or not
+     * @param role for which the check shall happen
+     * @return a Boolean indicating whether the user has the role or not
      */
-    fun isCurrentUserCompanyOwnerForCompany(companyId: String): Boolean {
+    fun doesCurrentUserHaveGivenRoleForCompany(companyId: String, role: CompanyRole): Boolean {
```
Comment (MichBlank) [inline dataland-backend/src/main/kotlin/org/dataland/datalandbackend/services/CompanyRoleChecker.kt:32]: maybe just hasCurrentUserRoleForCompany?

File: dataland-backend/src/main/kotlin/org/dataland/datalandbackend/services/CompanyRoleChecker.kt [hunk 3]
```diff
@@ -47,18 +47,43 @@ class CompanyOwnershipChecker(
     }
 
     /**
-     * Method to check whether the currently authenticated user is company owner of the specified company that is
+     * Checks whether the current user has any company role for the companyId
+     * @param companyId to check the roles for
+     * @returns a boolean stating if the user has any company role for the company
+     */
+    fun doesCurrentUserHaveAnyRoleForCompany(companyId: String): Boolean {
```
Comment (MichBlank) [inline dataland-backend/src/main/kotlin/org/dataland/datalandbackend/services/CompanyRoleChecker.kt:54]: similar to the previous one: hasCurrentUserAnyRoleForCompany?

File: dataland-e2etests/src/test/kotlin/org/dataland/e2etests/customApiControllers/CustomVsmeDataControllerApi.kt [hunk 4]
```diff
@@ -71,7 +72,8 @@ class CustomVsmeDataControllerApi(private val token: String) {
                 documents,
             ),
         ).execute()
-        require(response.isSuccessful) { "Request failed, response is: $response" }
+        if (response.code == 403) { throw ClientException("Client error : 403 ") }
```
Comment (MichBlank) [inline dataland-e2etests/src/test/kotlin/org/dataland/e2etests/customApiControllers/CustomVsmeDataControllerApi.kt:75]: Out of curiosity, why do we throw an exception with a manual message here and do not use the specific subclass?

File: dataland-e2etests/src/test/kotlin/org/dataland/e2etests/tests/communityManager/SingleDataRequestsTest.kt [hunk 5]
```diff
@@ -43,6 +45,8 @@ class SingleDataRequestsTest {
     val apiAccessor = ApiAccessor()
     val jwtHelper = JwtAuthenticationHelper()
     private val requestControllerApi = RequestControllerApi(BASE_PATH_TO_COMMUNITY_MANAGER)
+    private val maxRequestsForUser = 10
+    private val dataReaderUserId = UUID.fromString("18b67ecc-1176-4506-8414-1e81661017ca")
```
Comment (MichBlank) [inline dataland-e2etests/src/test/kotlin/org/dataland/e2etests/tests/communityManager/SingleDataRequestsTest.kt:49]: Why do we hard code the userId here and do not use TechnicalUser.Reader.technicalUserId as in the other cases in this PR?

File: dataland-frontend/src/components/pages/CompanyCockpitPage.vue [hunk 6]
```diff
@@ -107,8 +107,8 @@ export default defineComponent({
         | { [key in DataTypeEnum]: AggregatedFrameworkDataSummary }
         | undefined,
       FRAMEWORKS_WITH_VIEW_PAGE,
-      isUserCompanyOwner: false,
-      isUserUploader: false,
+      isUserCompanyOwnerOrUploader: false,
+      isUserDatalandUploader: false,
```
Comment (MichBlank) [inline dataland-frontend/src/components/pages/CompanyCockpitPage.vue:111]: maybe isUserKeycloakUploader to make the keycloak role more explicit?

File: dataland-frontend/src/components/pages/CompanyCockpitPage.vue [hunk 7]
```diff
@@ -137,7 +137,7 @@ export default defineComponent({
       if (!this.authenticated) {
         return false;
       }
-      return this.isUserCompanyOwner || this.isFrameworkPublic(framework);
+      return this.isUserCompanyOwnerOrUploader || this.isFrameworkPublic(framework);
```
Comment (emanueldfine) [inline dataland-frontend/src/components/pages/CompanyCockpitPage.vue:140]: Das ist falsch.  Jeder company-role Inhaber soll das dataset betrachten dürfen.

File: dataland-frontend/src/components/pages/CompanyCockpitPage.vue (hunk 6 above)
Comment (emanueldfine) [reply dataland-frontend/src/components/pages/CompanyCockpitPage.vue:111]: Ja stimmt ist besser. Habe es angepasst.

File: dataland-backend/src/main/kotlin/org/dataland/datalandbackend/services/CompanyRoleChecker.kt (hunk 2 above)
Comment (aliabdoli77) [reply dataland-backend/src/main/kotlin/org/dataland/datalandbackend/services/CompanyRoleChecker.kt:32]: good suggestion. applied

File: dataland-community-manager/src/main/kotlin/org/dataland/datalandcommunitymanager/repositories/CompanyRoleAssignmentRepository.kt [hunk 8]
```diff
@@ -28,4 +30,17 @@ interface CompanyRoleAssignmentRepository : JpaRepository<CompanyRoleAssignmentE
...
+     * @param userId to delete the assignments for
+     */
+    @Modifying
+    @Transactional
+    @Query(
+        "DELETE FROM CompanyRoleAssignmentEntity roleAssignment " +
+            "WHERE roleAssignment.companyId = :companyId " +
+            "AND roleAssignment.userId = :userId",
+    )
+    fun deleteAllByCompanyIdAndUserId(companyId: String, userId: String)
```
[the remark this answers is not in the document]
Comment (aliabdoli77) [reply dataland-community-manager/src/main/kotlin/org/dataland/datalandcommunitymanager/repositories/CompanyRoleAssignmentRepository.kt:45]: done

File: dataland-e2etests/src/test/kotlin/org/dataland/e2etests/tests/communityManager/SingleDataRequestsTest.kt (hunk 5 above)
Comment (aliabdoli77) [reply dataland-e2etests/src/test/kotlin/org/dataland/e2etests/tests/communityManager/SingleDataRequestsTest.kt:49]: done

File: dataland-e2etests/src/test/kotlin/org/dataland/e2etests/customApiControllers/CustomVsmeDataControllerApi.kt (hunk 4 above)
Comment (emanueldfine) [reply dataland-e2etests/src/test/kotlin/org/dataland/e2etests/customApiControllers/CustomVsmeDataControllerApi.kt:75]: Diesen client in diesem File musste ich selbst händisch schreiben, weil der autgenerierte client von openApi leider keinen vernünftigen multipart request machen kann.

An dieser Stelle hier möchte ich, dass mein custom client sich genauso verhält wie der autogenerierte client von openApi im Falle eines 403 errors.
Grund: Dann kann ich in den Tests auf dieselbe error message asserten wie bei den autogenerierten clients.

Daher "ahme" ich hier einfach das nach, was die autogenerierten clients auch machen.

File: dataland-e2etests/src/test/kotlin/org/dataland/e2etests/tests/communityManager/SingleDataRequestsTest.kt (hunk 5 above)
Comment (emanueldfine) [reply dataland-e2etests/src/test/kotlin/org/dataland/e2etests/tests/communityManager/SingleDataRequestsTest.kt:49]: Auf jeden Fall 👍 

gut gesehen

Habe es geändert

File: dataland-frontend/src/components/pages/CompanyCockpitPage.vue (hunk 7 above)
Comment (aliabdoli77) [reply dataland-frontend/src/components/pages/CompanyCockpitPage.vue]: corrected

File: dataland-backend/src/main/kotlin/org/dataland/datalandbackend/api/DataDeletionApi.kt (hunk 1 above)
Comment (aliabdoli77) [reply dataland-backend/src/main/kotlin/org/dataland/datalandbackend/api/DataDeletionApi.kt:37]: Answered by Andreas: No the uploader can delete all datasets from the company

Comment (aliabdoli77) [reply dataland-backend/src/main/kotlin/org/dataland/datalandbackend/api/DataDeletionApi.kt:37]: so its fine the way it is

File: dataland-e2etests/src/test/kotlin/org/dataland/e2etests/tests/CompanyRolesControllerTest.kt [hunk 9]
```diff
@@ -384,17 +384,51 @@ class CompanyRolesControllerTest {
...
+            ),
+            currentAssignments.first(),
+        )
+
+        assignCompanyRole(CompanyRole.Member, companyId, dataUploaderUserId)
+
+        currentAssignments = getCompanyRoleAssignments(companyId = companyId, userId = dataUploaderUserId)
+        assertEquals(1, currentAssignments.size)
+        assertEquals(
+            currentAssignments.first(),
```
Comment (MichBlank) [inline dataland-e2etests/src/test/kotlin/org/dataland/e2etests/tests/CompanyRolesControllerTest.kt:420]: One could think about switching the inputs as not to get confused what is actually expected when  the assertion fails. Could also be worthwile to write a small function inside of the method
```
fun assertRoleAssignments(role, currentAssignments)
```

**Pull request merged**